File: Word Ladder II.py
# Leetcode Problem 126: Word Ladder II
# PYTHON CODE
from collections import defaultdict, deque
from typing import List


class Solution:

    def findLadders(
        self, beginWord: str, endWord: str, wordList: List[str]
    ) -> List[List[str]]:
        word_set = set(wordList)
        if endWord not in word_set:
            return []

        # Step 1: BFS to find shortest distance from beginWord to all reachable words
        dist = {beginWord: 0}
        queue = deque([beginWord])
        found = False

        while queue and not found:
            level_size = len(queue)
            for _ in range(level_size):
                curr = queue.popleft()
                curr_dist = dist[curr]

                # Try changing every character from 'a' to 'z'
                for i in range(len(curr)):
                    original_char = curr[i]
                    for c in "abcdefghijklmnopqrstuvwxyz":
                        if c == original_char:
                            continue
                        next_word = curr[:i] + c + curr[i+1:]

                        if next_word in word_set:
                            if next_word not in dist:
                                dist[next_word] = curr_dist + 1
                                queue.append(next_word)
                                if next_word == endWord:
                                    found = True

        if endWord not in dist:
            return []

        # Step 2: DFS Backtracking from endWord to beginWord using distance map
        results = []

        def dfs(curr_word: str, path: List[str]):
            if curr_word == beginWord:
                results.append(path[::-1])  # Reverse path to start from beginWord
                return

            curr_dist = dist[curr_word]
            for i in range(len(curr_word)):
                original_char = curr_word[i]
                for c in "abcdefghijklmnopqrstuvwxyz":
                    if c == original_char:
                        continue
                    prev_word = curr_word[:i] + c + curr_word[i+1:]

                    # Only move to valid words in dist that are at (curr_dist - 1)
                    if prev_word in dist and dist[prev_word] == curr_dist - 1:
                        path.append(prev_word)
                        dfs(prev_word, path)
                        path.pop()

        dfs(endWord, [endWord])
        return results

File: Word Ladder II.java
// Leetcode Problem 126: Word Ladder II
// JAVA CODE
import java.util.*;

class Solution {
    public List<List<String>> findLadders(String beginWord, String endWord, List<String> wordList) {
        List<List<String>> results = new ArrayList<>();
        Set<String> wordSet = new HashSet<>(wordList);
        
        if (!wordSet.contains(endWord)) {
            return results;
        }

        // Phase 1: BFS to find the shortest distance from beginWord to all reachable words
        Map<String, Integer> dist = new HashMap<>();
        dist.put(beginWord, 0);
        
        Queue<String> queue = new LinkedList<>();
        queue.offer(beginWord);
        
        boolean found = false;
        int wordLen = beginWord.length();

        while (!queue.isEmpty() && !found) {
            int levelSize = queue.size();
            
            for (int i = 0; i < levelSize; i++) {
                String curr = queue.poll();
                int currDist = dist.get(curr);
                char[] charArray = curr.toCharArray();

                for (int j = 0; j < wordLen; j++) {
                    char originalChar = charArray[j];
                    
                    for (char c = 'a'; c <= 'z'; c++) {
                        if (c == originalChar) continue;
                        
                        charArray[j] = c;
                        String nextWord = new String(charArray);

                        if (wordSet.contains(nextWord)) {
                            if (!dist.containsKey(nextWord)) {
                                dist.put(nextWord, currDist + 1);
                                queue.offer(nextWord);
                                
                                if (nextWord.equals(endWord)) {
                                    found = true;
                                }
                            }
                        }
                    }
                    charArray[j] = originalChar; // Backtrack character
                }
            }
        }

        if (!dist.containsKey(endWord)) {
            return results;
        }

        // Phase 2: Backtracking DFS from endWord to beginWord
        List<String> path = new ArrayList<>();
        path.add(endWord);
        dfs(endWord, beginWord, dist, path, results, wordLen);

        return results;
    }

    private void dfs(String currWord, String beginWord, Map<String, Integer> dist, 
                     List<String> path, List<List<String>> results, int wordLen) {
        
        if (currWord.equals(beginWord)) {
            List<String> validPath = new ArrayList<>(path);
            Collections.reverse(validPath); // Reverse to start from beginWord
            results.add(validPath);
            return;
        }

        int currDist = dist.get(currWord);
        char[] charArray = currWord.toCharArray();

        for (int i = 0; i < wordLen; i++) {
            char originalChar = charArray[i];

            for (char c = 'a'; c <= 'z'; c++) {
                if (c == originalChar) continue;

                charArray[i] = c;
                String prevWord = new String(charArray);

                // Only step to valid predecessors with dist == currDist - 1
                if (dist.containsKey(prevWord) && dist.get(prevWord) == currDist - 1) {
                    path.add(prevWord);
                    dfs(prevWord, beginWord, dist, path, results, wordLen);
                    path.remove(path.size() - 1); // Backtrack
                }
            }
            charArray[i] = originalChar;
        }
    }
}